### DISCUSSION TITLE
Combining VerifyDirectory with System.IO.Abstractions

### DISCUSSION THREAD

**Discussion by michael-wolfenden** [VerifyTests/Verify#695] [Ideas]
[System.IO.Abstractions](https://github.com/TestableIO/System.IO.Abstractions) is an abstraction over System.IO that it extremely useful for testing IO.  At the core of the library is `IFileSystem`. Instead of calling methods like `File.ReadAllText` directly you use `IFileSystem.File.ReadAllText`.

The idea is that there would be an overload (or some other method) that allows you to call `VerifyDirectory` with the existing parameters but also passing an `IFileSystem`. This would allow you to build an entire mock file system in memory and then call `VerifyDirectory` to verify it's contents.

For example:

```csharp
[Fact]
public async Task UsingSystemIOAbstractions() {
    // MockFileSystem is provided by System.IO.Abstractions.TestingHelpers
    var fileSystem = new MockFileSystem(new Dictionary<string, MockFileData>
    {
        { @"c:\demo\jQuery.js", new MockFileData("some js") },
        { @"c:\demo\image.gif", new MockFileData(new byte[] { 0x12, 0x34, 0x56, 0xd2 }) },
        { @"c:\demo\subfolder\myfile.txt", new MockFileData("Testing is meh.") },
    });

    await VerifyDirectory(
        fileSystem,
        @"c:\demo",
        include: filePath => filePath.Contains("Doc"),
        pattern: "*.txt",
        options: new()
        {
            RecurseSubdirectories = false
        });
}
```

I imagine the implementation would be exactly the same as the existing implementation but replacing `Directory.EnumerateFiles` with `IFileSystem.Directory.EnumerateFiles` etc. I suppose `VerifyFile` could also be implemented as well.

I did attempt to implement myself however the existing implementation of `VerifyDirectory` relies heavily on a number of internal classes and methods.

**Comment by SimonCropp** [comment]
if i could make this (hypothetical) api work, would it suffice?

        var fileSystem = new MockFileSystem(new Dictionary<string, MockFileData>
        {
            { @"c:\demo\jQuery.js", new MockFileData("some js") },
            { @"c:\demo\image.gif", new MockFileData(new byte[] { 0x12, 0x34, 0x56, 0xd2 }) },
            { @"c:\demo\subfolder\myfile.txt", new MockFileData("Testing is meh.") },
        });

        var files = fileSystem.EnumerateFiles(
            @"c:\demo",
            "*.txt",
            SearchOption.AllDirectories);
        await Verify(files);

**Reply by michael-wolfenden** [reply]
I might be misunderstanding, but is your suggestion that Verify takes an enumerable of file paths?

If that's the case then it still wouldn't work as you need the `IFileSystem` to get the contents of the files via `fileSystem.File.ReadAllText(..)` for instance.

**Comment by SimonCropp** [comment]
i dont have capacity to add support System.IO.Abstractions. I am ahoy for you to create your own project that adds support
